File: Frontend/src/services/testDriveService.ts
import type { TestDrive, ApiResponse, PaginatedResponse } from '@/types';
import { TestDriveStatus } from '@/types';
import { apiRequest } from '@/lib/api';
import { normalizeTestDrive } from '@/lib/normalize';

class TestDriveService {
  async getTestDrives(
    page: number = 1,
    limit: number = 10,
    filters?: {
      status?: TestDriveStatus;
      customerId?: string;
      dealershipId?: string;
      dealerId?: string;
      upcoming?: boolean;
    }
  ): Promise<ApiResponse<PaginatedResponse<TestDrive>>> {
    const res = await apiRequest<PaginatedResponse<Record<string, unknown>>>('/test-drives', {
      params: { page, limit, status: filters?.status },
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to fetch test drives' };
    }

    let items = (res.data.data || []).map(normalizeTestDrive);

    if (filters?.customerId) items = items.filter(td => td.customerId === filters.customerId);
    if (filters?.dealershipId) items = items.filter(td => td.dealershipId === filters.dealershipId);
    if (filters?.dealerId) items = items.filter(td => td.assignedDealerId === filters.dealerId);
    if (filters?.upcoming) {
      const now = new Date();
      items = items.filter(td =>
        td.preferredDate >= now &&
        (td.status === TestDriveStatus.PENDING || td.status === TestDriveStatus.CONFIRMED)
      );
    }

    return {
      success: true,
      data: { ...res.data, data: items, total: items.length },
    };
  }

  async getTestDriveById(id: string): Promise<ApiResponse<TestDrive>> {
    const res = await apiRequest<Record<string, unknown>>(`/test-drives/${id}`);
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Test drive not found' };
    }
    return { success: true, data: normalizeTestDrive(res.data) };
  }

  async scheduleTestDrive(
    testDriveData: Omit<TestDrive, 'id' | 'status' | 'createdAt' | 'updatedAt'>
  ): Promise<ApiResponse<TestDrive>> {
    const date = testDriveData.preferredDate instanceof Date
      ? testDriveData.preferredDate.toISOString().split('T')[0]
      : String(testDriveData.preferredDate);

    const body = {
      customerId: Number(testDriveData.customerId),
      carId: Number(testDriveData.carId),
      dealershipId: Number(testDriveData.dealershipId),
      customerName: testDriveData.customerName,
      customerEmail: testDriveData.customerEmail,
      customerPhone: testDriveData.customerPhone,
      carModel: testDriveData.carModel,
      preferredDate: date,
      preferredTime: testDriveData.preferredTime,
      notes: testDriveData.notes,
    };

    const res = await apiRequest<Record<string, unknown>>('/test-drives', {
      method: 'POST',
      body: JSON.stringify(body),
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to schedule test drive' };
    }
    return { success: true, data: normalizeTestDrive(res.data), message: 'Test drive scheduled successfully' };
  }

  async updateTestDriveStatus(id: string, status: TestDriveStatus): Promise<ApiResponse<TestDrive>> {
    const res = await apiRequest<Record<string, unknown>>(`/test-drives/${id}/status`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to update status' };
    }
    return { success: true, data: normalizeTestDrive(res.data), message: `Test drive status updated to ${status}` };
  }

  async assignDealer(id: string, dealerId: string): Promise<ApiResponse<TestDrive>> {
    const res = await apiRequest<Record<string, unknown>>(`/test-drives/${id}/assign`, {
      method: 'PATCH',
      body: JSON.stringify({ dealerId: Number(dealerId) }),
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to assign dealer' };
    }
    return { success: true, data: normalizeTestDrive(res.data), message: 'Dealer assigned successfully' };
  }

  async addFeedback(id: string, feedback: string, rating: number): Promise<ApiResponse<TestDrive>> {
    const res = await apiRequest<Record<string, unknown>>(`/test-drives/${id}/feedback`, {
      method: 'PATCH',
      body: JSON.stringify({ feedback, rating }),
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to add feedback' };
    }
    return { success: true, data: normalizeTestDrive(res.data), message: 'Feedback added successfully' };
  }

  async getCustomerTestDrives(customerId: string): Promise<ApiResponse<TestDrive[]>> {
    const res = await apiRequest<Record<string, unknown>[]>(`/test-drives/customer/${customerId}`);
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to fetch test drives' };
    }
    return { success: true, data: res.data.map(normalizeTestDrive) };
  }

  async getUpcomingTestDrives(dealershipId?: string, limit: number = 5): Promise<ApiResponse<TestDrive[]>> {
    const res = await apiRequest<Record<string, unknown>[]>('/test-drives/upcoming', {
      params: { limit },
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to fetch upcoming test drives' };
    }
    return { success: true, data: res.data.map(normalizeTestDrive) };
  }

  async cancelTestDrive(id: string, reason?: string): Promise<ApiResponse<TestDrive>> {
    const res = await apiRequest<Record<string, unknown>>(`/test-drives/${id}/cancel`, {
      method: 'PATCH',
      body: JSON.stringify({ reason: reason || '' }),
    });
    if (!res.success || !res.data) {
      return { success: false, message: res.message || 'Failed to cancel test drive' };
    }
    return { success: true, data: normalizeTestDrive(res.data), message: 'Test drive cancelled successfully' };
  }
}

export const testDriveService = new TestDriveService();

File: Backend/src/main/java/com/serene/dms/service/TestDriveService.java
package com.serene.dms.service;

import java.time.LocalDate;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.serene.dms.dto.request.CreateTestDriveRequest;
import com.serene.dms.entity.Car;
import com.serene.dms.entity.Dealership;
import com.serene.dms.entity.TestDrive;
import com.serene.dms.entity.User;
import com.serene.dms.enums.TestDriveStatus;
import com.serene.dms.exception.ResourceNotFoundException;
import com.serene.dms.repository.CarRepository;
import com.serene.dms.repository.DealershipRepository;
import com.serene.dms.repository.TestDriveRepository;
import com.serene.dms.repository.UserRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
@RequiredArgsConstructor
public class TestDriveService {

    private final TestDriveRepository testDriveRepository;
    private final UserRepository userRepository;
    private final CarRepository carRepository;
    private final DealershipRepository dealershipRepository;

    public Page<TestDrive> getAll(int page, int limit, TestDriveStatus status) {
        PageRequest pageReq = PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.ASC, "preferredDate"));
        if (status != null) return testDriveRepository.findByStatus(status, pageReq);
        return testDriveRepository.findAll(pageReq);
    }

    public TestDrive getById(Long id) {
        return testDriveRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Test drive not found"));
    }

    @Transactional
    public TestDrive schedule(CreateTestDriveRequest request) {
        User customer;
        if (request.getCustomerId() != null) {
            customer = userRepository.findById(request.getCustomerId())
                    .orElseThrow(() -> new ResourceNotFoundException("Customer not found"));
        } else {
            customer = userRepository.findByEmail(request.getCustomerEmail())
                    .orElseGet(() -> {
                        String name = request.getCustomerName() != null ? request.getCustomerName() : "Guest";
                        int spaceIdx = name.indexOf(" ");
                        String fn = spaceIdx > 0 ? name.substring(0, spaceIdx) : name;
                        String ln = spaceIdx > 0 ? name.substring(spaceIdx + 1) : "";
                        User newUser = User.builder()
                                .email(request.getCustomerEmail())
                                .firstName(fn)
                                .lastName(ln)
                                .phone(request.getCustomerPhone())
                                .role(com.serene.dms.enums.UserRole.CUSTOMER)
                                .status(com.serene.dms.enums.UserStatus.ACTIVE)
                                .passwordHash("GUEST_" + java.util.UUID.randomUUID().toString())
                                .build();
                        return userRepository.save(newUser);
                    });
        }

        Car car = carRepository.findById(request.getCarId())
                .orElseThrow(() -> new ResourceNotFoundException("Car not found"));
        Dealership dealership = dealershipRepository.findById(request.getDealershipId())
                .orElseThrow(() -> new ResourceNotFoundException("Dealership not found"));

        TestDrive td = TestDrive.builder()
                .customer(customer)
                .customerName(request.getCustomerName())
                .customerEmail(request.getCustomerEmail())
                .customerPhone(request.getCustomerPhone())
                .car(car)
                .carModel(request.getCarModel())
                .dealership(dealership)
                .preferredDate(request.getPreferredDate())
                .preferredTime(request.getPreferredTime() != null ? request.getPreferredTime() : "")
                .notes(request.getNotes())
                .status(TestDriveStatus.PENDING)
                .build();
        TestDrive saved = testDriveRepository.save(td);
        log.info("Test drive scheduled: id={}, carId={}, date={}", saved.getId(), request.getCarId(), request.getPreferredDate());
        return saved;
    }

    @Transactional
    public TestDrive updateStatus(Long id, TestDriveStatus status) {
        TestDrive td = getById(id);
        td.setStatus(status);
        log.info("Test drive status updated: id={}, status={}", id, status);
        return testDriveRepository.save(td);
    }

    @Transactional
    public TestDrive assignDealer(Long id, Long dealerId) {
        TestDrive td = getById(id);
        User dealer = userRepository.findById(dealerId)
                .orElseThrow(() -> new ResourceNotFoundException("Dealer not found"));
        td.setAssignedDealer(dealer);
        return testDriveRepository.save(td);
    }

    @Transactional
    public TestDrive addFeedback(Long id, String feedback, Integer rating) {
        TestDrive td = getById(id);
        td.setFeedback(feedback);
        td.setRating(rating);
        return testDriveRepository.save(td);
    }

    @Transactional
    public TestDrive cancel(Long id, String reason) {
        TestDrive td = getById(id);
        td.setStatus(TestDriveStatus.CANCELLED);
        if (reason != null) td.setNotes(reason);
        return testDriveRepository.save(td);
    }

    public List<TestDrive> getByCustomer(Long customerId) {
        return testDriveRepository.findByCustomerIdOrderByPreferredDateDesc(customerId);
    }

    public List<TestDrive> getUpcoming(int limit) {
        List<TestDriveStatus> active = List.of(TestDriveStatus.PENDING, TestDriveStatus.CONFIRMED);
        return testDriveRepository.findByStatusInAndPreferredDateGreaterThanEqualOrderByPreferredDateAsc(
                active, LocalDate.now()).stream().limit(limit).toList();
    }
}
